src/adapters/philips_hue/mod.rs

```diff
@@ -0,0 +1,284 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+extern crate serde_json;
+
+mod http;
+mod nupnp;
+mod api;
+
+use adapters::philips_hue::api::structs::*;
+use adapters::philips_hue::api::HueLight;
+use controller::Controller;
+use events::*;
+use iron::{ Request, Response, IronResult };
+use iron::headers::{ ContentType, AccessControlAllowOrigin };
+use iron::status::Status;
+use iron::method::Method;
+use router::Router;
+use service::{ Service, ServiceAdapter, ServiceProperties };
+use std::time::Duration;
+use std::thread;
+use uuid::Uuid;
+use std::io::Read;
+
+
+pub struct PhilipsHueAdapter<T> {
+    name: String,
+    controller: T,
+}
+
+impl<T: Controller> PhilipsHueAdapter<T> {
+    pub fn new(controller: T) -> Self {
+        debug!("Creating Philips Hue adapter");
+        PhilipsHueAdapter { name: "PhilipsHueAdapter".to_owned(),
+                       controller: controller,
+                     }
+    }
+}
+
+impl<T: Controller> ServiceAdapter for PhilipsHueAdapter<T> {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn start(&self) {
+        let mut id = 0;
+        let controller = self.controller.clone();
+
+        thread::spawn(move || {
+            controller.send_event(
+                EventData::AdapterStart {
+                    name: "Philips Hue Service Adapter".to_owned()
+                }).unwrap();
+
+            let nupnp_hubs = nupnp::query();
+            debug!("nUPnP reported Philips Hue bridges: {:?}", nupnp_hubs);
+
+            let available_hubs = nupnp_hubs
+                .iter()
+                .filter(|hub| hub.is_available())
+                .collect::<Vec<_>>();
+
+            let paired_hubs = available_hubs
+                .iter()
+                .filter(|hub| {
+                    info!("Philips Hue bridge {} at {}", hub.id, hub.ip);
+
+                    // Querying the settings fails when not paired with the bridge
+                    if hub.is_paired() {
+                        true
+                    } else {
+                        controller.send_event(
+                            EventData::AdapterNotification {
+                                info: format!(
+                                    "Push pairing button on Philips Hue bridge with ID {}",
+                                    hub.id)
```

Comment:
Let's turn that into a json snippet, as simple as: { "adapter": "philips_hue", "message": "NeedsPairing", "hub": "$hub_id" }
And the UI will manage the l10n part and copy-writing